base/usr/share/man/man1/man.1
.TH MAN 1 "March, 2026"

.SH NAME
\fBman\fP, \fBapropos\fP, \fBwhatis\fP \- Find and display manual pages.

.SH SYNOPSIS
\fBman\fP [\fB\-a\fP] [\fB\-w\fP] [\fB\-S\fP \fIsections\fP] [\fIsection\fP] \fIpage\fP...
.sp
\fBman\fP \fB\-k\fP [\fB\-S\fP \fIsections\fP] \fIkeyword\fP...
.br
\fBapropos\fP [\fB\-s\fP \fIsections\fP] \fIkeyword\fP...
.sp
\fBman\fP \fB\-f\fP [\fB\-S\fP \fIsections\fP] \fIpage\fP...
.br
\fBwhatis\fP [\fB\-s\fP \fIsections\fP] \fIpage\fP...

.SH DESCRIPTION
\fBman\fP looks for man pages in \fB/usr/share/man/man\f(BIsection\fB/\f(BIpage\fB.\f(BIsection\fR, where
\fIsection\fP is a section number and \fIpage\fP is the page to look for.
.PP
If there is more than one argument, and the first argument is a single character, or begins with a digit followed
by only lowercase letters, it is interpreted as \fIsection\fP. Otherwise, it is assumed to be another \fIpage\fP.
.PP
If no \fIsection\fP is specified, then sections specified either by the \fB\-S\fP option, the \fBMANSECT\fP
environment variable, or the default set of \fB"1:n:l:8:3:0:2:3krk:5:4:9:6:7"\fP will be searched in order
and the first result, if any, will be shown. If no manual page is found by the requested name, a diagnostic
message is printed.
.PP
If a \fIpage\fP argument looks like it might be a file path (eg. it contains a /), lookup is skipped and
the file path is tried directly.
.PP
The manual page is displayed by piping \fIroff\fP to \fImore -rP'page(section)' --stay --alt\fP.
.PP
If a \fIpage\fP argument that was interpreted as a file path ends in \fB.gz\fP, or a manual page located
through the section lookup described above ends in \fB.gz\fP, it will be decompressed
by \fIgunzip\fP before being passed to \fIroff\fP.
.PP
If multiple \fIpage\fP arguments are provided, each will be searched and displayed in sequence.

.SS Options
.TP
.B \-a \fRor\fP \-\-all
Display all matching manual pages, in search order, one after the other, instead of stopping after
showing the first match.

.TP
.B \-w \fRor\fP \-\-where
Instead of displaying the manual page, print its location.

.TP
.B \-S \fIsections\fP \fRor\fP \-\-sections=\fIsections\fP
Specify the \fIsections\fP to look in as a colon\-separated list. This option overrides any value
retrieved from the \fBMANSECT\fP environment variable.
Lowercase \fB\-s\fP is also accepted for compatibility.

.TP
.B \-k \fRor\fP \-\-apropos
Search the available manuals for pages whose \fBNAME\fP section contains \fIkeyword\fP.
This is a wrapper around \fBroff \-S NAME \-P | grep -i\fP.
.sp
Each \fIkeyword\fP will be searched separately.
Any \fIkeyword\fP for which no matching manual was found will result in a diagnostic message.
There is no cached database of page summaries, so this search may take a moment.

.TP
.B \-f \fRor\fP \-\-whatis
Display just the first line of the \fBNAME\fP section, similar to the output from \fB\-k\fP, but
only for exact matches of manual pages, rather than searching the content.
This option, or invoking \fBman\fP as \fBwhatis\fP, implies the \fB\-\-all\fP option.

.SH EXAMPLES

.TP
.B man man
View this manual page.

.TP
.B man -k yutani
See a list of manual pages related to the windowing system.

.TP
.B man -a basename
View the manual pages for the
.IR basename (1)
utility and then the
.IR basename (3)
library function, in sequence.

.TP
.B man 3 basename
View the manual page for the
.IR basename (3)
library function (instead of viewing the manual page for the utility).

.TP
.B man -w man
Locate the file that provides the source for this manual page.

.TP
.B man /usr/share/man/man1/man.1
View this manual page directly from its source.

.SH NOTES
Be warned that \fBman\fP often passes arguments to
.IR system (3)
without any input validation.
It is the responsibility of the user to not shell inject themselves with a bad input.
In particular, \fIpage\fP and \fIkeyword\fP arguments will both be passed to various
utilities in single quotes without any additional escaping.

.SH EXIT STATUS
If any of the requested \fIpage\fP arguments could not be found, or another error was encountered,
then \fIman\fP will exit with status 1.

.SH ENVIRONMENT
.TP
.B MANSECT
A colon\-separated list of sections to search in order, which overrides the default.

.SH FILES
The search path for manual pages is \fB/usr/share/man\fP.
This implementation will only look for manuals in that directory.
.PP
Only a limited \fBroff\fP dialect is supported at this time.

.SH STANDARDS
\fBman\fP is described in the User Portability section of POSIX.1 in limited detail.
Only the \fB\-k\fP option is required, and its specific behavior is left largely to the implementation.
We believe our \fBman\fP conforms sufficiently to the specification, though our definition
of "sufficiently" may not meet certification requirements.
.PP
The \fB\-a\fP, \fB\-w\fP, and \fB\-S\fP options generally match their respective behavior in most other implementations.

.SH SEE ALSO
.BR roff (1),
.BR more (1)

.SH HOMEPAGE
.PP
.B ToaruOS:
https://toaruos.org/

.SH AUTHOR
This \fBman\fP was written by K. Lange <klange@toaruos.org>.


apps/man.c
/**
 * @brief Rudimentary manpage viewer.
 *
 * Searches for a manual page at /usr/share/man/manN/PAGE.N,
 * and if one is found it is pasesed to a 'roff|more' pipeline
 * for display.
 *
 * @copyright
 * This file is part of ToaruOS and is released under the terms
 * of the NCSA / University of Illinois License - see LICENSE.md
 * Copyright (C) 2026 K. Lange
 */
#include <stdio.h>
#include <stdlib.h>
#include <termios.h>
#include <unistd.h>
#include <string.h>
#include <stdint.h>
#include <dirent.h>
#include <errno.h>
#include <getopt.h>
#include <libgen.h>
#include <sys/wait.h>
#include <sys/stat.h>

#if !defined(TOARU_MAN_TOOL_PREFIX)
#define TOARU_MAN_TOOL_PREFIX ""
#endif

#if !defined(TOARU_MAN_PATH_PREFIX)
#define TOARU_MAN_PATH_PREFIX ""
#endif

#define MAN_DIR  TOARU_MAN_PATH_PREFIX "/usr/share/man/man%s"
#define MAN_FMT  MAN_DIR "/%s.%s"
#define ROFF_CMD TOARU_MAN_TOOL_PREFIX "roff"
#define MORE_CMD TOARU_MAN_TOOL_PREFIX "more -rP'%s(%s)' --stay --alt"

/* Don't bother prefixing this; I don't want to build our grep elsewhere. */
#define GREP_CMD "grep --color -i -- '%s' -"

static int where_is = 0;
static int what_is = 0;

static int show_help(char * argv[]) {
#define X_S "\033[3m"
#define X_E "\033[0m"
	fprintf(stderr,
		"usage: %s [-a] [-S " X_S "sectionlist" X_E "] [" X_S "section" X_E "] " X_S "page..." X_E "\n"
		"       %s -k [-S " X_S "sectionlist" X_E "] " X_S "keyword..." X_E "\n"
		"\n"
		"Display a manual page.\n"
		"\n"
		"Options:\n"
		"\n"
		" -a       " X_S "Display all matching manuals." X_E "\n"
		" -w       " X_S "Print the locations of manual pages." X_E "\n"
		" -S " X_S "list  Limit search to colon-separated list of sections." X_E "\n"
		" -k       " X_S "Search manual page NAME sections for keywords." X_E "\n"
		" -f       " X_S "Display a short description from each matching manual." X_E "\n"
		"          " X_S "This options implies " X_E "-a.\n"
		" --help   " X_S "Show this help text." X_E "\n"
		"\n", argv[0], argv[0]);
	return 0;
}

static int usage(char * argv[]) {
	fprintf(stderr, "Try '%s --help' or '%s man' for more information.\n", argv[0], argv[0]);
	return 1;
}

static void print_page_name(const char * page, const char * i) {
	int written = fprintf(stdout, "%s (%s)", page, i);
	while (written < 20) {
		fprintf(stdout, " ");
		written++;
	}
	fprintf(stdout, " - ");
	fflush(stdout);
}

/**
 * @brief Try to display a manual page.
 *
 * Checks for the existence of @p filename and, if found, possibly
 * decompresses it, passes it to @b roff, and pipes it to @b more.
 *
 * @param filename Full path to the file to check.
 * @param page Page name to display in the @b more prompt.
 * @param i Section number to display in the @b more prompt.
 * @returns 1 if the stat succeeded and we ran the pipeline (even if the pipeline fails)
 *          otherwise 0 if the stat failed.
 */
static int try_filename(char * filename, char * page, char * i) {
	struct stat st;
	if (!stat(filename, &st)) {
		if (where_is) {
			puts(filename);
			return 1;
		}
		int is_gz = strlen(filename) > 3 && !strcmp(filename + strlen(filename)-3,".gz");
		char * systemcmd;
		if (what_is) {
			print_page_name(page, i);
			asprintf(&systemcmd,
				!is_gz ?
				(ROFF_CMD " -S NAME -P -I1 '%s' | head -n1") :
				("gunzip -c '%s' | " ROFF_CMD " -S NAME -P -I1 -- - | head -n1"),
				filename);
		} else {
			asprintf(&systemcmd,
				!is_gz ?
				(ROFF_CMD " '%s' | " MORE_CMD) :
				("gunzip -c '%s' | " ROFF_CMD " -- - | " MORE_CMD),
				filename, page, i);
		}
		int result = system(systemcmd);
		if (result) exit(WEXITSTATUS(result));
		free(systemcmd);
		return 1;
	}
	return 0;
}

/**
 * @brief Look for a page in a section.
 *
 * Checks whether @p page in section @p i exists
 * at the configured location, with or without
 * a @b .gz suffix.
 *
 * @param i Section to look in.
 * @param page Page name to look for.
 * @returns 0 if neither filename worked, 1 if one did.
 */
static int try_section(char *i, char * page) {
	char * filename;
	asprintf(&filename, MAN_FMT, i, page, i);
	if (try_filename(filename, page, i)) {
		free(filename);
		return 1;
	}
	free(filename);
	asprintf(&filename, MAN_FMT ".gz", i, page, i);
	if (try_filename(filename, page, i)) {
		free(filename);
		return 1;
	}
	free(filename);
	return 0;
}

/**
 * @brief Search pages in a section for a keyword.
 *
 * Looks at all the manual pages in section @p i and processes
 * them through 'roff -S NAME -P | grep -i' to determine if
 * they contain a requested @p keyword. First, this is done
 * in a dry-run mode where the output is written to /dev/null.
 * Then, if the page matched, its name is printed and it is
 * processed again for actual output.
 *
 * This will only try to look at files in the section that
 * actually look like pages, so if you ask for section 1
 * and a file doesn't match '{page}.1' or '{page}.1.gz',
 * it will be ignored. This matches the behavior of our page
 * lookup elsewhere, as man will not find these pages if
 * ask for them by name.
 *
 * We try to pass @p keyword to grep as a single argument,
 * but nothing in this stack does input validation, so it's
 * up to the user to not shell inject themselves with a bad
 * keyword argument.
 *
 * Because we run 'roff' with the '-P' option, grep should
 * be able to match across formatting changes correctly.
 *
 * Also we enable --color in grep because it looks nice.
 *
 * @param i Section to search in
 * @param keyword Pattern to pass to grep
 * @returns 1 if a page matched or 0 otherwise.
 */
static int search_section(char * i, char * keyword) {
	size_t len_i = strlen(i);
	char * dirpath;
	asprintf(&dirpath, MAN_DIR, i);
	DIR * dir = opendir(dirpath);
	if (!dir) return free(dirpath), 0; /* Not a valid section? */

	int found = 0;
	struct dirent * ent;
	while ((ent = readdir(dir))) {
		if (ent->d_name[0] == '.') continue;

		char * filename;
		asprintf(&filename,"%s/%s", dirpath, ent->d_name);
		int is_gz = strlen(filename) > 3 && !strcmp(filename + strlen(filename)-3,".gz");

		char * page = strdup(ent->d_name);
		size_t len_page = strlen(page);

		if (is_gz) {
			/* Exclude .gz suffix from printed name */
			page[len_page - 3] = '\0';
			len_page -= 3;
		}

		if (len_page > len_i + 1 && page[len_page - len_i - 1] == '.' &&
			!strcmp(page + len_page - len_i, i)) {
			/* Exclude section suffix from printed name */
			page[len_page - len_i - 1] = '\0';
			len_page -= len_i + 1;
		} else {
			/* And reject files that don't look like man pages for this section. */
			free(page);
			free(filename);
			continue;
		}

		int dry_run = 1;
		while (dry_run >= 0) {

			if (!dry_run) {
				/* If not in dry run, prefix the output without the page and section
				 * so the user can actually find it. Try to format this nicely. */
				print_page_name(page, i);
			}

			char * systemcmd;
			asprintf(&systemcmd,
				!is_gz ?
				(ROFF_CMD " -S NAME -P -I1 '%s' | " GREP_CMD "%s") :
				("gunzip -c '%s' | " ROFF_CMD " -S NAME -P -I1 -- - | " GREP_CMD "%s"),
				filename, keyword, dry_run ? " >/dev/null" : "");
			int result = system(systemcmd);
			free(systemcmd);
			if (result != 0) break;

			found |= 1;
			dry_run--;
		}

		free(page);
		free(filename);
	}
	free(dirpath);
	closedir(dir);
	return found;
}

/**
 * @brief Convert a colon-separated list into an array.
 *
 * Parses a colon-separated list of strings into an array of those
 * strings. Used to parse the @c MANSECT environment variable or
 * the @b -S option argument.
 *
 * @param str String to convert.
 * @returns An allocated array of pointers to strings in a copy of @c str
 */
static char ** sections_from_string(char * str) {
	size_t count = 1;
	char * c = str;
	for (; *c; ++c) {
		if (*c == ':') count++;
	}
	char ** sects = calloc(count + 1, sizeof(char*));
	size_t i = 0;
	c = str;
	while (*c) {
		sects[i] = c;
		while (*c && *c != ':') c++;
		if (*c == ':') {
			*c = '\0';
			i++;
			c++;
		}
	}

	return sects;
}

/**
 * @brief Check if a string looks like a filename.
 *
 * Just checks if a path contains a /
 *
 * @param arg Possible filename.
 * @returns 1 if it looks like a filename, else 0.
 */
static int is_file_name(char * arg) {
	while (*arg) {
		if (*arg == '/') return 1;
		arg++;
	}
	return 0;
}

/**
 * @brief Check if a string looks like a section number.
 *
 * A section number can be a single lowercase letter,
 * such as @b n or @b l or it can be one numeric digit
 * followed by any amount of lower case letters. Is that
 * technically correct? Probably not, but it's close enough
 * to reality.
 *
 * @param arg Possible section number.
 * @returns 1 if it looks like a section, else 0.
 */
static int is_section(char * arg) {
	if (*arg >= 'a' && *arg <= 'z' && !arg[1]) return 1;
	if (*arg >= '0' && *arg <= '9') {
		arg++;
		while (*arg) {
			if (*arg < 'a' || *arg > 'z') return 0;
			arg++;
		}
		return 1;
	}
	return 0;
}

int main(int argc, char * argv[]) {

	/* The default section list either comes from @b $MANSECT - but only
	 * if we're building for ToaruOS and it is set - or the default
	 * list that we stole from an Ubuntu configuration. */
	char * section_list =
#if defined(__toaru__)
		getenv("MANSECT");
	if (!section_list) section_list =
#endif
			"1:n:l:8:3:0:2:3krk:5:4:9:6:7";

	section_list = strdup(section_list);

	int show_all = 0;
	int search_names = 0;
	int opt;

	if (argc && !strcmp(basename(argv[0]), TOARU_MAN_TOOL_PREFIX "whatis")) what_is = 1;
	if (argc && !strcmp(basename(argv[0]), TOARU_MAN_TOOL_PREFIX "apropos")) search_names = 1;

	static struct option long_opts[] = {
		{"apropos",  no_argument, 0, 'k'},
		{"all",      no_argument, 0, 'a'},
		{"where",    no_argument, 0, 'w'},
		{"path",     no_argument, 0, 'w'},
		{"location", no_argument, 0, 'w'},
		{"whatis",   no_argument, 0, 'f'},
		{"sections", required_argument, 0, 'S'},
		{"help",     no_argument, 0, 1000},
		{0,0,0,0}
	};

	while ((opt = getopt_long(argc, argv, "awS:s:kf", long_opts, NULL)) != -1) {
		switch (opt) {
			case 'a':
				show_all = 1;
				break;
			case 'w':
				where_is = 1;
				break;
			case 's':
			case 'S':
				section_list = optarg;
				break;
			case 'k':
				search_names = 1;
				break;
			case 'f':
				what_is = 1;
				break;
			case 1000:
				return show_help(argv);
			case '?':
				return usage(argv);
		}
	}

	if (what_is) show_all = 1;

	/* At least one argument is required. */
	if (optind == argc) {
		return 0;
	}

	/* Convert the default sections list (from the envvar, compiled default, or argument)
	 * to an array we can iterate over more easily. */
	char ** sections = sections_from_string(section_list);
	char * section = NULL;
	int ret = 0;

	/* If there is more than one argument and it looks like it might be a section
	 * then treat it is as such and use it for all of the rest of the arguments. */
	if ((!search_names && !what_is) && optind + 1 != argc && is_section(argv[optind])) {
		section = argv[optind];
		optind++;
	}

	for (; optind != argc; ++optind) {
		int found = 0;

		if (search_names) {
			for (char ** s = section ? (char*[]){section, NULL} : sections; *s; ++s) {
				found |= search_section(*s, argv[optind]);
			}
		} else if (is_file_name(argv[optind])) {
			/* Accept things that look like raw paths to roff sources */
			if (try_filename(argv[optind], argv[optind], "")) found = 1;
		} else if (section) {
			/* If a single section argument was given, look in that. */
			if (try_section(section, argv[optind])) found = 1;
		} else {
			/* Otherwise, search through the MANSECT list in order. */
			for (char ** s = sections; *s; ++s) {
				if (try_section(*s, argv[optind])) {
					found = 1;
					if (!show_all) break;
				}
			}
		}

		if (!found) {
			fprintf(stderr,
				(search_names || what_is)
					? "%s: nothing appropriate\n"
					: "No manual entry for %s\n",
				argv[optind]);
			ret |= 1;
		}
	}

	free(section_list);
	free(sections);
	return ret;
}
